### PR TITLE
Add property "interface_type" and improve networks

### PR DESCRIPTION
**Pull request by dataclouder**
* Add property "interface_type" to routed network, allowing the choice
  among "internal" (default), "subinterface", and "distributed" [#321]
* Add property "description" in all three network resources.
* Add validation for IP fields in routed and isolated networks. [#373] [#374]
* Update documentation

Note for testing:
To test distributed networking, you need to have an edge gateway with distributed routing enabled, and then you need to enable such tests by setting the variable `VCD_TEST_DISTRIBUTED_NETWORK` to a non-empty value.

### CODE REVIEW

File: vcd/resource_vcd_network_routed.go [hunk 1]
```diff
@@ -444,26 +511,42 @@ func resourceVcdNetworkIPAddressHash(v interface{}) int {
 // On success, returns the name of the edge gateway
 func findEdgeGatewayConnection(client *VCDClient, vdc *govcd.Vdc, network *govcd.OrgVDCNetwork) (string, error) {
```
Comment (lvirbalas) [inline vcd/resource_vcd_network_routed.go]: The name of this function is very confusing. I was reading the big comment at:
https://github.com/terraform-providers/terraform-provider-vcd/pull/402/files#diff-e99542a876fc291b70cf88bd368423b5R41-R48

It just didn't make any sense, then I looked at this function and realized it's not returning a "Connection". Better name needed.

File: vcd/datasource_vcd_network_test.go [hunk 2]
```diff
@@ -38,6 +38,14 @@ func detectNetwork(vcdClient *VCDClient, vdc *govcd.Vdc, network *govcd.OrgVDCNe
...
 	}
 	edgeGatewayName, err := findEdgeGatewayConnection(vcdClient, vdc, network)
+	// There is a potential false positive returning from findEdgeGatewayConnection:
+	// When a non-routed network has the same name as the edge gateway,
+	// this function will return as if it had found an edge gateway.
+	// This fact won't affect regular operations, as this function is called
+	// from functions that assume a routed network. The only place where
+	// it could cause a problem is in the data source tests (this file), where this function
+	// is used to find the network type. If we avoid calling both the edge gateway and the
+	// isolated network "foo", all will be well.
```
Comment (lvirbalas) [inline vcd/datasource_vcd_network_test.go]: Why not to add this comment to the function itself?
But first, why is this happening, how come network is even considered together with a totally different type of EG?

File: vcd/resource_vcd_network_routed.go (hunk 1 above)
Comment (dataclouder) [reply vcd/resource_vcd_network_routed.go]: The name says what the function is doing: finding a connection between a routed network and an edge gateway. Please suggest a better name if you have one.

File: vcd/datasource_vcd_network_test.go (hunk 2 above)
Comment (dataclouder) [reply vcd/datasource_vcd_network_test.go]: The comment is here because it may only affect this kind of test. Putting it in the function would raise unnecessary questions about risks that don't exist for regular usage.

File: vcd/resource_vcd_network_routed.go (hunk 1 above)
Comment (dataclouder) [reply vcd/resource_vcd_network_routed.go]: More background info:
when we create a routed network, we know the name of the edge gateway. When we retrieve a network for importing into a state file, we don't know the edge gateway name, as it is not needed to retrieve the network. Thus, once we have the network, we need to find to which edge gateway is connected. This function exists since the creation of the network data source. It was using a different algorithm then, by looking at the existing edge gateways for a connected network. However, for distributed networks the edge gateway doesn't store a connection, and we had to change the method of connection retrieval using a query.

Comment (lvirbalas) [reply vcd/resource_vcd_network_routed.go]: Makes it clearer. Assuming I got the purpose right, then the name suggestion is:
`findEdgeGatewayNameByNetwork`
This would be clearer and consistent with our format of `GetEdgeGatewayByName`

Finally, I just realized that this function is in provider, we should really move it to govcd.

Comment (lvirbalas) [reply vcd/resource_vcd_network_routed.go]: One more variant:
`findEdgeGatewayNameOfNetwork`

**Comment by dataclouder** [comment]
`findGatewayConnection` moved to go-vcloud-director as `vdc.FindEdgeGatewayNameByNetwork`

File: vcd/datasource_vcd_network_test.go [hunk 3]
```diff
@@ -285,6 +296,10 @@ output "network_name" {
   value = data.vcd_network_direct.{{.NetworkName}}.name
 }
 
+output "network_description" {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_network_test.go]: I did that for a specific reason: using output fields will allow me to use the generated HCL script to see the result when running it with `terraform` tool. If I don't put them, I will have to edit the file to see the result.  Since I had to do this operation several times when I was developing, I thought that I get the best of two worlds with this approach: result in the integrated tests, and visible output when running manually.

File: CHANGELOG.md [hunk 4]
```diff
@@ -31,6 +31,12 @@ IMPROVEMENTS:
 * Add command `make token` to create an authorization token from testing credentials
 * Clean up interpolation-only expressions from tests (as allowed in terraform v0.12.11+)
 * Increment vCD API version used from 27.0 to 29.0 [GH-396]
+* `resource/vcd_network_routed` Add properties `description` and `interface` [GH-321] [GH-342] [GH-374]
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply CHANGELOG.md]: fixed

File: vcd/resource_vcd_network_routed.go (hunk 1 above)
Comment (dataclouder) [reply vcd/resource_vcd_network_routed.go]: Moved to govcd

File: vcd/datasource_vcd_network_test.go (hunk 2 above)
Comment (dataclouder) [reply vcd/datasource_vcd_network_test.go]: removed comment and refactored as the interested function was moved to govcd

File: vcd/resource_vcd_network_routed.go [hunk 5]
```diff
@@ -214,6 +246,27 @@ func resourceVcdNetworkRoutedCreate(d *schema.ResourceData, meta interface{}) er
...
+	case "internal":
+		// default: no configuration is needed
+		orgVDCNetwork.Configuration.SubInterface = nil
+	case "subinterface":
+		orgVDCNetwork.Configuration.SubInterface = &trueValue
+	case "distributed":
+		if distributedAllowed {
+			orgVDCNetwork.Configuration.DistributedInterface = &trueValue
+		} else {
+			return fmt.Errorf("interface 'distributed' requested, but edge gateway '%s' not enabled", edgeGateway.EdgeGateway.Name)
```
Comment (Didainius) [inline vcd/resource_vcd_network_routed.go]: I'd think this could be rephrased to better. Currently I read the error - "edge gateway is not enabled" while in fact "distributed routing is not enabled on edge gateway"

Comment (dataclouder) [reply vcd/resource_vcd_network_routed.go]: Thanks. Changed

File: vcd/resource_vcd_network_routed.go [hunk 6]
```diff
@@ -50,35 +50,56 @@ func resourceVcdNetworkRouted() *schema.Resource {
...
+			},
+
+			"interface_type": &schema.Schema{
+				Type:             schema.TypeString,
+				Optional:         true,
+				Default:          "internal",
+				ForceNew:         true,
+				Description:      "Which interface to use (one of `internal`, `subinterface`, `distributed`)",
+				ValidateFunc:     validation.StringInSlice([]string{"internal", "subinterface", "distributed"}, true),
+				DiffSuppressFunc: suppressNetworkUpgradedInterface(),
```
Comment (vbauzysvmware) [inline vcd/resource_vcd_network_routed.go]: As far as I got it it's support back compatibility. So comment would be good regarding that here
